File: android/app/src/main/kotlin/com/stockbook/app/feature/settings/BackupScreen.kt
package com.stockbook.app.feature.settings

import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.imePadding
import androidx.compose.foundation.layout.statusBarsPadding
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.stockbook.app.design.GhostButton
import com.stockbook.app.design.Glyph
import com.stockbook.app.design.Icon
import com.stockbook.app.design.Metrics
import com.stockbook.app.design.Nocturne
import com.stockbook.app.design.NocturneType
import com.stockbook.app.design.PrimaryButton
import com.stockbook.app.design.ScreenHeader
import com.stockbook.app.design.SecondaryButton
import com.stockbook.app.design.card
import com.stockbook.app.design.hairline
import com.stockbook.app.photos.PhotoStore
import com.stockbook.core.model.ShopState
import com.stockbook.core.store.StockbookStore
import com.stockbook.core.text.Strings
import com.stockbook.core.transfer.BackupArchive

/**
 * The export/import handoff — the app's only route onto a new phone.
 *
 * Both directions go through the system document picker, so the file lands
 * wherever the owner keeps things and the app never asks for storage permission
 * to do it. Import **validates before asking anything**, and only then offers to
 * replace the database.
 */
@Composable
fun BackupScreen(
    state: ShopState,
    store: StockbookStore,
    strings: Strings,
    onClose: () -> Unit
) {
    val context = LocalContext.current
    val importFlow = remember { ImportFlow() }

    val exporter = rememberLauncherForActivityResult(
        ActivityResultContracts.CreateDocument("application/zip")
    ) { uri ->
        // Only a real write counts. A cancelled save sheet must not claim a
        // backup exists — that is the one lie this screen must never tell.
        if (uri == null) return@rememberLauncherForActivityResult
        val photos = PhotoStore(context)
        runCatching {
            context.contentResolver.openOutputStream(uri)?.use { out ->
                // The pictures are read one at a time as the writer asks for
                // them, so a shop with two hundred of them never holds more than
                // one. A picture this phone has lost is skipped rather than
                // fatal, the same way the sweep leaves its id alone.
                out.write(BackupArchive.pack(store.makeBackupDocument()) { photos.bytes(it) })
            }
        }.onSuccess { store.markExported() }
    }

    val importer = rememberLauncherForActivityResult(
        ActivityResultContracts.OpenDocument()
    ) { uri -> importFlow.pick(uri, context.contentResolver) }

    // Drawn as a sibling of the tab content inside the shell's Box, so without a
    // ground of its own the screen behind shows straight through it — and without
    // `statusBarsPadding` its header sits under the status bar.
    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(Nocturne.bg)
            .statusBarsPadding()
            .imePadding()
    ) {
        ScreenHeader(title = strings.moveToAnotherPhone) {
            GhostButton(strings.done, onClick = onClose)
        }

        Column(
            modifier = Modifier
                .fillMaxWidth()
                .verticalScroll(rememberScrollState())
                .padding(horizontal = Metrics.screenPadding)
                .padding(bottom = 18.dp)
        ) {
            Text(
                strings.moveToAnotherPhoneNote,
                style = NocturneType.inter(12.5),
                color = Nocturne.neutral500,
                modifier = Modifier.padding(bottom = 8.dp)
            )

            // This used to be an accent-coloured warning that the pictures
            // stayed behind. They travel now, so it is a plain fact rather than
            // something to be careful about — and worth saying at all only
            // because a photograph of a paper bill is the thing an owner would
            // most expect a backup to miss.
            Text(
                strings.photosTravelWithTheBook,
                style = NocturneType.inter(12.5),
                color = Nocturne.neutral500,
                modifier = Modifier.padding(bottom = 14.dp)
            )

            // Export
            Column(modifier = Modifier.fillMaxWidth().card().padding(13.dp)) {
                CardHeading(Icon.confirm, strings.exportEverything)
                Text(
                    if (state.settings.hasBackup) strings.exportNoteAfterBackup else strings.exportNoteFirstTime,
                    style = NocturneType.inter(12.0),
                    color = Nocturne.neutral500,
                    modifier = Modifier.padding(bottom = 11.dp)
                )
                PrimaryButton(
                    title = if (state.settings.hasBackup) strings.writeAFreshFile else strings.createBackupFile,
                    onClick = { exporter.launch(store.makeBackupDocument().suggestedFilename) },
                    fullWidth = true,
                    height = 42.dp,
                    fontSize = 13.5
                )
            }

            Spacer(Modifier.height(10.dp))

            // Import
            Column(modifier = Modifier.fillMaxWidth().card().padding(13.dp)) {
                CardHeading(Icon.openRow, strings.importABackupFile)

                val current = importFlow.stage
                Text(
                    when (current) {
                        is ImportFlow.Stage.Imported -> strings.importNoteDone
                        is ImportFlow.Stage.Failed -> strings.backupError(current.error)
                        else -> strings.importNoteIdle
                    },
                    style = NocturneType.inter(12.0),
                    color = if (current is ImportFlow.Stage.Failed) Nocturne.accent300 else Nocturne.neutral500,
                    modifier = Modifier.padding(bottom = 11.dp)
                )

                if (current is ImportFlow.Stage.Picked) {
                    Column(
                        modifier = Modifier
                            .fillMaxWidth()
                            .hairline(Nocturne.accent, Metrics.controlRadius)
                            .padding(11.dp)
                    ) {
                        Text(
                            current.filename,
                            style = NocturneType.inter(13.0),
                            color = Nocturne.text,
                            maxLines = 1,
                            overflow = TextOverflow.Ellipsis
                        )
                        Text(
                            current.document.summaryLine(strings),
                            style = NocturneType.inter(11.5),
                            color = Nocturne.neutral500,
                            modifier = Modifier.padding(top = 7.dp)
                        )
                        // Naming what is about to be lost, in the owner's own
                        // numbers. The last thing between a tap and an
                        // unrecoverable swap.
                        Text(
                            strings.replaceWarning(state.products.size, state.bills.size),
                            style = NocturneType.inter(11.5),
                            color = Nocturne.accent300,
                            modifier = Modifier.padding(top = 8.dp)
                        )
                    }
                    Spacer(Modifier.height(10.dp))
                    Row(modifier = Modifier.fillMaxWidth()) {
                        SecondaryButton(
                            strings.cancel,
                            onClick = { importFlow.cancel() },
                            fullWidth = true,
                            height = 42.dp,
                            fontSize = 13.5,
                            modifier = Modifier.weight(1f)
                        )
                        Spacer(Modifier.width(8.dp))
                        PrimaryButton(
                            strings.replaceEverything,
                            onClick = {
                                val confirmed = importFlow.confirm() ?: return@PrimaryButton
                                store.replaceEverything(confirmed)
                                val photos = PhotoStore(context)
                                // The book first, then its pictures out of the
                                // same archive, then the sweep. In that order:
                                // the sweep keeps what the *current* book names,
                                // so anything left over from the replaced book
                                // goes and everything just restored stays.
                                importFlow.restorePhotos(context.contentResolver, photos)
                                photos.sweep(store.photoIdsInUse())
                            },
                            fullWidth = true,
                            height = 42.dp,
                            fontSize = 13.5,
                            modifier = Modifier.weight(1f)
                        )
                    }
                } else {
                    SecondaryButton(
                        strings.chooseAFile,
                        // `*/*` because the document picker is unreliable about
                        // types — a `.zip` arrives as `application/octet-stream`
                        // often enough that filtering on the name would hide the
                        // owner's own backup from them. The bytes are what get
                        // checked.
                        onClick = { importer.launch(arrayOf("application/zip", "application/json", "*/*")) },
                        fullWidth = true,
                        height = 42.dp,
                        fontSize = 13.5
                    )
                }
            }
        }
    }
}

@Composable
private fun CardHeading(icon: androidx.compose.ui.graphics.vector.ImageVector, title: String) {
    Row(verticalAlignment = Alignment.CenterVertically, modifier = Modifier.padding(bottom = 8.dp)) {
        Glyph(icon, size = 17.dp, tint = Nocturne.accent)
        Spacer(Modifier.width(9.dp))
        Text(title, style = NocturneType.rowPrimary, color = Nocturne.text)
    }
}

File: ios/Stockbook/Features/Settings/BackupScreen.swift
import SwiftUI
import UniformTypeIdentifiers

/// The export/import handoff — the app's only route onto a new phone.
///
/// Lifted out of Settings, which had grown into one long scroll where the two
/// things an owner changes weekly sat above the two they use once a year. This
/// is the once-a-year half, behind a row.
///
/// The prototype faked the file layer. This does not: export writes a real dated
/// JSON file through the document exporter, Share hands the same file to the OS
/// share sheet, and import reads a file the owner picked, **validates it before
/// asking anything**, and only then offers to replace the database.
struct BackupScreen: View {
    @Environment(StockbookStore.self) private var store
    @Environment(AppRouter.self) private var router

    // Export
    @State private var exportDocument: BackupFile?
    @State private var isExporting = false
    @State private var exportChip: (name: String, detail: String)?
    @State private var shareURL: URL?
    @State private var seeded = false

    // Import
    @State private var isImporting = false
    @State private var importFlow = ImportFlow()

    private var settings: Settings { store.settings }
    private var products: [Product] { store.products }
    private var bills: [Bill] { store.bills }

    var body: some View {
        VStack(spacing: 0) {
            ScreenHeader(title: Loc.moveToAnotherPhone) {
                Button(Loc.done) { router.showingBackup = false }
                    .buttonStyle(GhostButtonStyle(fontSize: 12.5))
            }

            ScrollView {
                VStack(alignment: .leading, spacing: 0) {
                    Text(Loc.moveToAnotherPhoneNote)
                        .font(NocturneType.inter(12.5))
                        .foregroundStyle(Nocturne.neutral500)
                        .lineSpacing(3)
                        .padding(.bottom, 14)

                    exportCard.padding(.bottom, 10)
                    importCard
                }
                .padding(.horizontal, Metrics.screenPadding)
                .padding(.bottom, 18)
            }
        }
        .background(Nocturne.bg.ignoresSafeArea())
        .ignoresSafeArea(.keyboard, edges: .bottom)
        .onAppear(perform: seed)
        .fileExporter(
            isPresented: $isExporting,
            document: exportDocument,
            contentType: .json,
            defaultFilename: exportDocument?.document.suggestedFilename
        ) { result in
            // Only a real write counts. A cancelled save sheet must not claim a
            // backup exists — that is the one lie this screen must never tell.
            if case .success = result {
                store.markExported()
                refreshExportChip()
            }
        }
        .fileImporter(
            isPresented: $isImporting,
            allowedContentTypes: [.json]
        ) { result in
            importFlow.pick(result)
        }
    }

    // MARK: Export

    private var exportCard: some View {
        VStack(alignment: .leading, spacing: 0) {
            cardHeading(icon: Icon.export, title: Loc.exportEverything)

            Text(exportNote)
                .font(NocturneType.inter(12))
                .foregroundStyle(Nocturne.neutral500)
                .lineSpacing(3)
                .padding(.bottom, 11)

            if let chip = exportChip {
                HStack(spacing: 10) {
                    Glyph(Icon.file, size: 19)
                        .foregroundStyle(Nocturne.accent400)
                    VStack(alignment: .leading, spacing: 1) {
                        Text(chip.name)
                            .font(NocturneType.inter(13))
                            .lineLimit(1)
                            .truncationMode(.middle)
                        Text(chip.detail)
                            .font(NocturneType.inter(11))
                            .foregroundStyle(Nocturne.neutral500)
                    }
                    Spacer(minLength: 0)
                }
                .padding(.horizontal, 11)
                .padding(.vertical, 10)
                .background(Nocturne.bg, in: RoundedRectangle(cornerRadius: Metrics.controlRadius, style: .continuous))
                .hairline(radius: Metrics.controlRadius)
                .padding(.bottom, 10)
            }

            HStack(spacing: 8) {
                Button(settings.hasBackup ? Loc.writeAFreshFile : Loc.createBackupFile) {
                    exportDocument = BackupFile(document: store.makeBackupDocument())
                    isExporting = true
                }
                .buttonStyle(PrimaryButtonStyle(fullWidth: true, height: 42, fontSize: 13.5))

                if let shareURL {
                    ShareLink(item: shareURL) {
                        Label(Loc.share, systemImage: Icon.share)
                    }
                    .buttonStyle(SecondaryButtonStyle(height: 42, fontSize: 13.5))
                }
            }
        }
        .padding(13)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(Nocturne.surface, in: RoundedRectangle(cornerRadius: Metrics.cardRadius, style: .continuous))
    }

    // MARK: Import

    private var importCard: some View {
        VStack(alignment: .leading, spacing: 0) {
            cardHeading(icon: Icon.importFile, title: Loc.importABackupFile)

            Text(importNote)
                .font(NocturneType.inter(12))
                .foregroundStyle(importFlow.stage.isFailure ? Nocturne.accent300 : Nocturne.neutral500)
                .lineSpacing(3)
                .padding(.bottom, 11)

            if case .picked(let document, let filename) = importFlow.stage {
                VStack(alignment: .leading, spacing: 0) {
                    Text(filename)
                        .font(NocturneType.inter(13))
                        .lineLimit(1)
                        .truncationMode(.middle)
                        .padding(.bottom, 7)

                    Text(document.summaryLine(Loc))
                        .font(NocturneType.inter(11.5))
                        .foregroundStyle(Nocturne.neutral500)
                        .lineSpacing(4)

                    // Naming what is about to be lost, in the owner's own
                    // numbers. This is the last thing standing between a tap and
                    // an unrecoverable swap.
                    Text(Loc.replaceWarning(productCount: products.count, billCount: bills.count))
                        .font(NocturneType.inter(11.5))
                        .foregroundStyle(Nocturne.accent300)
                        .lineSpacing(3)
                        .padding(.top, 8)
                }
                .frame(maxWidth: .infinity, alignment: .leading)
                .padding(11)
                .background(Nocturne.bg, in: RoundedRectangle(cornerRadius: Metrics.controlRadius, style: .continuous))
                .hairline(Nocturne.accent, radius: Metrics.controlRadius)
                .padding(.bottom, 10)

                HStack(spacing: 8) {
                    Button(Loc.cancel) { importFlow.cancel() }
                        .buttonStyle(SecondaryButtonStyle(fullWidth: true, height: 42, fontSize: 13.5))
                    Button(Loc.replaceEverything) {
                        // Only ever acts on what confirm() hands back.
                        guard let confirmed = importFlow.confirm() else { return }
                        store.replaceEverything(with: confirmed)
                        // A swap, not a merge: every photograph this phone was
                        // holding belonged to the book that was just replaced.
                        // Only ids the incoming book names survive — and an id
                        // it names that this phone lacks is left waiting, not
                        // tidied away.
                        PhotoStore().sweep(keeping: store.photoIDsInUse())
                        refreshExportChip()
                    }
                    .buttonStyle(PrimaryButtonStyle(fullWidth: true, height: 42, fontSize: 13.5))
                }
            } else {
                Button { isImporting = true } label: {
                    Label(Loc.chooseAFile, systemImage: Icon.folder)
                }
                .buttonStyle(SecondaryButtonStyle(fullWidth: true, height: 42, fontSize: 13.5))
            }
        }
        .padding(13)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(Nocturne.surface, in: RoundedRectangle(cornerRadius: Metrics.cardRadius, style: .continuous))
    }

    private func cardHeading(icon: String, title: String) -> some View {
        HStack(spacing: 9) {
            Glyph(icon, size: 17).foregroundStyle(Nocturne.accent)
            Text(title).nocturneText(.rowPrimary)
            Spacer(minLength: 0)
        }
        .padding(.bottom, 4)
    }

    // MARK: Copy

    private var exportNote: String {
        settings.hasBackup ? Loc.exportNoteAfterBackup : Loc.exportNoteFirstTime
    }

    private var importNote: String {
        switch importFlow.stage {
        case .imported: Loc.importNoteDone
        case .failed(let error): Loc.backupError(error)
        default: Loc.importNoteIdle
        }
    }

    // MARK: Actions

    private func seed() {
        guard !seeded else { return }
        seeded = true
        refreshExportChip()
    }

    /// Encoding the whole database is cheap at this size but not free, so the
    /// chip is computed on demand rather than in `body`.
    private func refreshExportChip() {
        guard let exportedAt = settings.lastExportAt else {
            exportChip = nil
            shareURL = nil
            return
        }
        let document = store.makeBackupDocument(at: exportedAt)
        exportChip = (
            name: document.suggestedFilename,
            detail: [
                Loc.products(document.products.count),
                Loc.bills(document.bills.count),
                Loc.fileSize(kilobytes: BackupService.sizeInKilobytes(of: document))
            ].joined(separator: " · ")
        )
        // Share hands the OS a real file, so one has to exist on disk.
        shareURL = try? BackupService.writeToTemporaryFile(document)
    }
}
